File: src/Cli.Tests/Internal/DockerTests.cs
using System.Text;
using Docker.DotNet;
using Google.Protobuf;
using Xunit.Abstractions;

namespace UnMango.Tdl.Cli.Tests.Internal;

public sealed class DockerTests(ITestOutputHelper test)
{
	[Fact]
	public async Task TestDocker() {
		using var client = new DockerClientConfiguration()
			.CreateClient();

		var docker = new Cli.Internal.Docker(new XUnitConsole(test), client, "uml2ts");
		var spec = new Spec {
			Name = "test-spec",
			Types_ = {
				["testType"] = new Type {
					Type_ = "object",
				},
			},
		};

		await using MemoryStream input = new(spec.ToByteArray()), output = new();
		Assert.True(input.Length > 0, $"input length was: {input.Length}");

		await docker.GenerateAsync(input, output);

		Assert.True(output.Length > 0, $"output length was: {output.Length}");
		var actual = Encoding.UTF8.GetString(output.ToArray());
		test.WriteLine(actual);
		Assert.Equal("export interface testType \n{\n}", actual);
	}
}

File: src/FSharp.Tests/DockerTests.fs
module DockerTests

open System
open System.IO
open System.Text
open Docker.DotNet
open FsCheck.FSharp
open FsCheck.Xunit
open UnMango.Tdl
open Xunit

type Message =
  static member Double() =
    ArbMap.defaults
    |> ArbMap.arbitrary<string>
    |> Arb.filter (fun s -> s.Length > 0 && not <| s.Contains('\\'))

let rand = Random()

let testWorkload = Docker.Workload.create $"tdl-test-{rand.Next()}"

[<Property(Arbitrary = [| typeof<Message> |], Skip = "This is all sorts of broken")>]
let ``Can perform IO with a container`` (message: string) = async {
  use input = new MemoryStream(Encoding.UTF8.GetBytes(message), false)
  use output = new MemoryStream()
  use config = new DockerClientConfiguration()
  use client = config.CreateClient()

  let workload =
    { testWorkload "ubuntu" "latest" with
        Entrypoint = [ "cat" ] }

  do! client |> Docker.exec workload input output
  let actual = Encoding.UTF8.GetString(output.ToArray())

  Assert.True(output.Length > 0)
  Assert.Equal(message, actual)
}
